File: 1_data_preprocessing/self_data_preprocessing_template.R
"
#This tutorial is for Data preprocessing 

Original dataset:
    Country	Age	   Salary	Purchased
    France	 44	   72000	No
    Spain	   27	   48000	Yes
    Germany	 30	   54000	No
    Spain	   38	   61000	No
    Germany	 40		        Yes
    France	 35	   58000	Yes
    Spain		       52000	No
    France	 48	   79000	Yes
    Germany	 50	   83000	No
    France	 37	   67000	Yes
"

#step 1: import the dataset
#set a working directory by going to the Files pane, find the correct folder, then go
#to More and click on "Set as Working Directory"
dataset = read.csv('Data.csv')
#dataset = dataset[, 2:3]


#Step 2: Splitting the dataset into the Training set and Test set
#install caTools library by install.packages('caTools')
#Then, check caTools in the Packages pane. OR use
library(caTools)
set.seed(123) #set seed so that every time we get the same random result

#sample.split() will randomly select specified percentage of tuples to be the training
#set and label them as TRUE. The test set tuples are labeled as FALSE
split = sample.split(dataset$Purchased, SplitRatio = 0.8)

#create traning set and test set based on the TRUE and FALSE label created in the 
#last step in dataset
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

#Feature scaling
# training_set[, 2:3] = scale(training_set[, 2:3])
# test_set[, 2:3] = scale(test_set[, 2:3])

File: 1_data_preprocessing/self_data_preprocessing_template.py
"""
    This is for Data Preprocessing
    
    
Original dataset:
    Country	Age	Salary	Purchased
    France	   44	   72000	No
    Spain	   27	   48000	Yes
    Germany	30	   54000	No
    Spain	   38	   61000	No
    Germany	40		        Yes
    France	   35	   58000	Yes
    Spain		      52000	No
    France	   48	   79000	Yes
    Germany	50	   83000	No
    France	   37	   67000	Yes

"""

"""
Step 1: importing the libraries
"""

import numpy as np
import matplotlib.pyplot as plt

#panda's library is the best library to import and manage datasets 
import pandas as pd

"""
Step 2: importing the dataset
"""
#2.1: set a working directory by saving this file to the desired working
#directory. The folder that holds this file will the be working directory
#header = 0 means reads data WITH the header
dataset = pd.read_csv('Data.csv', header = 0)

#Replace some special characters in the header by '_' so that these headers
#can be read correctly in step 6 below
#dataset.columns = [x.strip().replace(' ', '_') for x in dataset.columns]
#dataset.columns = [x.strip().replace('&', '_') for x in dataset.columns]

#Print the headers
#print(list(dataset))

#2.2: create a matrix of features/IVs
X = dataset.iloc[:,:-1].values #[:,:-1] means select all the rows and the 
                               #columns except the last column

#2.3: create a dependent variable vector
y = dataset.iloc[:,3].values


"""
Step 3: Splitting the dataset into the training and test sets
"""
from sklearn.cross_validation import train_test_split
#random_state = 0 is to make everytime it gives the same result. it is like
#specifying a seed for random() in java
#practically, we do not need this so that every time the program will
#randomly select 20% of data for the test.
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

"""
Step 4: Show statistical results of the model built on the training set
"""
from statsmodels.formula.api import ols

#split the dataset to training and test sets
train_set, test_set = train_test_split(dataset, test_size = 0.2, random_state = 0)

#create the formula string
tempHeaders = list(dataset)
formula = tempHeaders[-1] + " ~ "
for iv in range(0, len(tempHeaders) - 1):
    formula += tempHeaders[iv] + " + "

formula = formula[:-3]

#use ols to create the model
model = ols(formula, train_set).fit()
print(model.summary())



"""
Step 4: Feature Scaling: Standardization and Normalization

Standardization equation:
   Xstand = (X - mean(X)) / (standard deviation(X))

Normalization equation:
   Xnorm = (X - min(X)) / (max(X) - min(X))

Note: most ML libraries deal with feature scaling. But some do not


#import StandardScaler 
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
"""







